### PR TITLE
create /healthcheck and /metrics endpoints in workers

### LINKED ISSUE #2067: Instrument the workers to profile jobs duration
The workers do not send metrics to Prometheus, which makes it hard to understand where the time is spent.

### PR DESCRIPTION
**Pull request by severo**
fixes #2067

- [x] Publish /metrics and /healthcheck endpoints in workers
- [x] Check worker's health with the new endpoint
- [x] Fetch worker's Prometheus metrics
- [x] Instrument parts of the worker's code

### CODE REVIEW

**Comment by severo** [comment]
Note that the default metrics can be interesting to debug the workers, cc @huggingface/datasets-server 

```
# HELP python_gc_objects_collected_total Objects collected during gc
# TYPE python_gc_objects_collected_total counter
python_gc_objects_collected_total{generation="0"} 4892.0
python_gc_objects_collected_total{generation="1"} 791.0
python_gc_objects_collected_total{generation="2"} 11.0
# HELP python_gc_objects_uncollectable_total Uncollectable object found during GC
# TYPE python_gc_objects_uncollectable_total counter
python_gc_objects_uncollectable_total{generation="0"} 0.0
python_gc_objects_uncollectable_total{generation="1"} 0.0
python_gc_objects_uncollectable_total{generation="2"} 0.0
# HELP python_gc_collections_total Number of times this generation was collected
# TYPE python_gc_collections_total counter
python_gc_collections_total{generation="0"} 377.0
python_gc_collections_total{generation="1"} 34.0
python_gc_collections_total{generation="2"} 3.0
# HELP python_info Python platform information
# TYPE python_info gauge
python_info{implementation="CPython",major="3",minor="9",patchlevel="15",version="3.9.15"} 1.0
# HELP process_virtual_memory_bytes Virtual memory size in bytes.
# TYPE process_virtual_memory_bytes gauge
process_virtual_memory_bytes 8.29612032e+08
# HELP process_resident_memory_bytes Resident memory size in bytes.
# TYPE process_resident_memory_bytes gauge
process_resident_memory_bytes 1.75685632e+08
# HELP process_start_time_seconds Start time of the process since unix epoch in seconds.
# TYPE process_start_time_seconds gauge
process_start_time_seconds 1.69935634806e+09
# HELP process_cpu_seconds_total Total user and system CPU time spent in seconds.
# TYPE process_cpu_seconds_total counter
process_cpu_seconds_total 3.09
# HELP process_open_fds Number of open file descriptors.
# TYPE process_open_fds gauge
process_open_fds 13.0
# HELP process_max_fds Maximum number of open file descriptors.
# TYPE process_max_fds gauge
process_max_fds 1.048576e+06
```

In addition to them, we can create new metrics. The first one I'll use is `method_steps_processing_time_seconds` to get an idea of the time passed in the most important methods

```
# HELP method_steps_processing_time_seconds Histogram of the processing time of specific steps in methods for a given context (in seconds)
# TYPE method_steps_processing_time_seconds histogram
```

**Comment by lhoestq** [comment]
Maybe the executor can start the app instead of the main worker loop ?

**Comment by severo** [comment]
> Nice ! Maybe the executor can start the app instead of the main worker loop ?

Yes, I was looking into this, and indeed it's not at the right place

**Comment by severo** [comment]
edit:

<strike>Note that the default metrics can be interesting to debug the workers, cc https://github.com/orgs/huggingface/teams/datasets-server</strike>

https://github.com/huggingface/datasets-server/pull/2069#issuecomment-1798323632

These metrics are not available, after all, because they don't work with multiprocess apps (see https://github.com/prometheus/client_python#multiprocess-mode-eg-gunicorn).

**Comment by severo** [comment]
It works. An extract of the worker /metrics:

```
# HELP method_steps_processing_time_seconds Multiprocess metric
# TYPE method_steps_processing_time_seconds histogram
method_steps_processing_time_seconds_sum{context="None",method="loop",step="start_job"} 0.7544647829927271
method_steps_processing_time_seconds_sum{context="None",method="loop",step="sleep"} 10.898265495998203
method_steps_processing_time_seconds_sum{context="dataset-config-names",method="loop",step="run_job"} 2.096199264997267
method_steps_processing_time_seconds_sum{context="num_jobs_to_create=8",method="CreateJobsTask.run",step="all"} 0.03710693099128548
method_steps_processing_time_seconds_sum{context="None",method="loop",step="finish_job"} 0.43577435999759473
method_steps_processing_time_seconds_sum{context="config-split-names-from-streaming",method="loop",step="run_job"} 1.5823528829932911
method_steps_processing_time_seconds_sum{context="num_jobs_to_create=5",method="CreateJobsTask.run",step="all"} 0.020024819998070598
method_steps_processing_time_seconds_sum{context="config-parquet-and-info",method="loop",step="run_job"} 5.527396860008594
method_steps_processing_time_seconds_sum{context="num_jobs_to_create=3",method="CreateJobsTask.run",step="all"} 0.014728451002156362
method_steps_processing_time_seconds_sum{context="dataset-parquet",method="loop",step="run_job"} 0.004233769999700598
method_steps_processing_time_seconds_sum{context="dataset-info",method="loop",step="run_job"} 0.005180924999876879
method_steps_processing_time_seconds_sum{context="dataset-size",method="loop",step="run_job"} 0.004504704003920779
method_steps_processing_time_seconds_sum{context="num_jobs_to_create=1",method="CreateJobsTask.run",step="all"} 0.0059098970086779445
method_steps_processing_time_seconds_sum{context="dataset-split-names",method="loop",step="run_job"} 0.005446648996439762
method_steps_processing_time_seconds_sum{context="dataset-is-valid",method="loop",step="run_job"} 0.004613705008523539
method_steps_processing_time_seconds_sum{context="dataset-opt-in-out-urls-count",method="loop",step="run_job"} 0.004844259994570166
method_steps_processing_time_seconds_bucket{context="None",le="0.005",method="loop",step="start_job"} 0.0
method_steps_processing_time_seconds_bucket{context="None",le="0.01",method="loop",step="start_job"} 0.0
method_steps_processing_time_seconds_bucket{context="None",le="0.025",method="loop",step="start_job"} 7.0
method_steps_processing_time_seconds_bucket{context="None",le="0.05",method="loop",step="start_job"} 18.0
method_steps_processing_time_seconds_bucket{context="None",le="0.075",method="loop",step="start_job"} 20.0
method_steps_processing_time_seconds_bucket{context="None",le="0.1",method="loop",step="start_job"} 20.0
method_steps_processing_time_seconds_bucket{context="None",le="0.25",method="loop",step="start_job"} 21.0
method_steps_processing_time_seconds_bucket{context="None",le="0.5",method="loop",step="start_job"} 21.0
method_steps_processing_time_seconds_bucket{context="None",le="0.75",method="loop",step="start_job"} 21.0
method_steps_processing_time_seconds_bucket{context="None",le="1.0",method="loop",step="start_job"} 21.0
method_steps_processing_time_seconds_bucket{context="None",le="2.5",method="loop",step="start_job"} 21.0
method_steps_processing_time_seconds_bucket{context="None",le="5.0",method="loop",step="start_job"} 21.0
method_steps_processing_time_seconds_bucket{context="None",le="7.5",method="loop",step="start_job"} 21.0
method_steps_processing_time_seconds_bucket{context="None",le="10.0",method="loop",step="start_job"} 21.0
method_steps_processing_time_seconds_bucket{context="None",le="+Inf",method="loop",step="start_job"} 21.0
method_steps_processing_time_seconds_count{context="None",method="loop",step="start_job"} 21.0
method_steps_processing_time_seconds_bucket{context="None",le="0.005",method="loop",step="sleep"} 0.0
...
```

**Comment by severo** [comment]
There is an error in the worker unit tests, surely due to my misunderstanding of how asyncio works

https://github.com/huggingface/datasets-server/actions/runs/6785400458/job/18443632155?pr=2069

File: services/worker/src/worker/app.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+            Route("/healthcheck", endpoint=healthcheck_endpoint),
+            Route("/metrics", endpoint=create_metrics_endpoint()),
+        ]
+    )
+
+
+async def start_web_app() -> None:
+    uvicorn_config = UvicornConfig.from_env()
+    config = uvicorn.Config(
+        "app:create_app",
```
Comment (lhoestq) [inline services/worker/src/worker/app.py:25]: maybe try this ?

```suggestion
        "worker.app:create_app",
```

Comment (lhoestq) [reply services/worker/src/worker/app.py:25]: It should fix the test that is erroring with
```
ModuleNotFoundError: No module named 'app'
```

File: Makefile [hunk 2]
```diff
@@ -23,19 +24,19 @@ include tools/Docker.mk
 
 .PHONY: start
 start:
-	MONGO_PORT=${MONGO_PORT} ADMIN_UVICORN_PORT=${PORT_ADMIN} API_UVICORN_PORT=${PORT_API} ROWS_UVICORN_PORT=${PORT_ROWS} SEARCH_UVICORN_PORT=${PORT_SEARCH} SSE_API_UVICORN_PORT=${PORT_SSE_API} PORT_REVERSE_PROXY=${PORT_REVERSE_PROXY} DOCKER_COMPOSE=${DOCKER_COMPOSE} $(MAKE) up
+	MONGO_PORT=${MONGO_PORT} ADMIN_UVICORN_PORT=${PORT_ADMIN} API_UVICORN_PORT=${PORT_API} ROWS_UVICORN_PORT=${PORT_ROWS} SEARCH_UVICORN_PORT=${PORT_SEARCH} SSE_API_UVICORN_PORT=${PORT_SSE_API} WORKER_UVICORN_PORT=${PORT_SSE_API} PORT_REVERSE_PROXY=${PORT_REVERSE_PROXY} DOCKER_COMPOSE=${DOCKER_COMPOSE} $(MAKE) up
```
Comment (AndreaFrancis) [inline Makefile:27]:
```suggestion
	MONGO_PORT=${MONGO_PORT} ADMIN_UVICORN_PORT=${PORT_ADMIN} API_UVICORN_PORT=${PORT_API} ROWS_UVICORN_PORT=${PORT_ROWS} SEARCH_UVICORN_PORT=${PORT_SEARCH} SSE_API_UVICORN_PORT=${PORT_SSE_API} WORKER_UVICORN_PORT=${PORT_WORKER} PORT_REVERSE_PROXY=${PORT_REVERSE_PROXY} DOCKER_COMPOSE=${DOCKER_COMPOSE} $(MAKE) up
```

Comment (AndreaFrancis) [reply Makefile:27]: Same for others?

Comment (severo) [reply Makefile:27]: Sure! Thanks

**Review by AndreaFrancis — APPROVED** [review]

**Review by rtrompier — APPROVED** [review]

**Comment by lhoestq** [comment]
> I was wondering @lhoestq : is it still useful to pass a command parameter (--print-worker-state-path). It's explicit, which is good, but as we always set it in the code, we might just remove this test, and always print the banner, no?

No strong opinion. I'd keep it so that we can run the worker loop bare-bone if we want. This is specific to the executor's logic.

**Comment by severo** [comment]
oh, it's still failing... https://github.com/huggingface/datasets-server/pull/2069

It was working locally. Looking at it

**Comment by severo** [comment]
The test error is random... Reminds me of https://github.com/huggingface/datasets-server/pull/1567

**Comment by severo** [comment]
makes sense, because the tasks that kill jobs are run in executor.py with a random interval, which might explain randomness in test results.

**Comment by severo** [comment]
OK, hopefully it's OK now

**Comment by severo** [comment]
Let's try! Merging

**Pull request merged**